lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -302,26 +372,67 @@ void SessionStateEnforcer::m5g_complete_termination(
-  // remove all rules;
-  MLOG(MINFO) << "Will be removing all associated rules of session id "
+void SessionStateEnforcer::m5g_state_change_action(
+    std::string& imsi, SessionStateUpdateCriteria& session_uc,
+    SetSmNotificationContext notif, std::unique_ptr<SessionState>& session) {
+  /*Irrespective of any State of Session, release and terminate*/
+  if ((notif.common_context().sm_session_state() == ACTIVE_2) &&
+      (session->get_state() == INACTIVE)) {
+    m5g_execute_state_change_action(session_uc, session, ACTIVE);
+    /* Call for all rules to be associated from session
+     * and inform to UPF
+     */
+    m5g_pdr_rules_change_and_update_upf(imsi, session, PdrState::INSTALL);
+  }
+  /* Now code for Inactive state notificaiton
+   */
+  else if (
+      (notif.common_context().sm_session_state() == INACTIVE_3) &&
+      (session->get_state() == ACTIVE)) {
+    m5g_execute_state_change_action(session_uc, session, INACTIVE);
+    /* Call for all rules to be de-associated from session
+     * and inform to UPF
+     */
+    m5g_pdr_rules_change_and_update_upf(imsi, session, PdrState::IDLE);
+  }
+}
+
+void SessionStateEnforcer::m5g_execute_state_change_action(
+    SessionStateUpdateCriteria& session_uc,
+    std::unique_ptr<SessionState>& session, SessionFsmState targetState) {
+  auto stateStr = session_fsm_state_to_str(targetState);
+  MLOG(MINFO) << stateStr << " state transition for session id "
               << session->get_session_id();
-  auto ip_addr = session->get_config()
-                     .rat_specific_context.m5gsm_session_context()
-                     .pdu_address()
-                     .redirect_server_address();
+  session->set_fsm_state(targetState, session_uc);
+  uint32_t cur_version = session->get_current_version();
+  session->set_current_version(++cur_version, session_uc);
+  MLOG(MINFO) << "During " << stateStr << " state of session changed to "
+              << session_fsm_state_to_str(session->get_state());
+  return;
+}
+
+void SessionStateEnforcer::m5g_pdr_rules_change_and_update_upf(
+    const std::string& imsi, const std::unique_ptr<SessionState>& session,
+    enum PdrState pdrstate) {
+  session->set_all_pdrs(pdrstate);
+  session->reset_throttle();
+  m5g_send_session_request_to_upf(imsi, session);
+  return;
+}
+
+void SessionStateEnforcer::m5g_send_session_request_to_upf(
+    const std::string& imsi, const std::unique_ptr<SessionState>& session) {
   // Update to UPF
   SessionState::SessionInfo sess_info;
-  sess_info.imsi         = imsi;
-  sess_info.ip_addr      = ip_addr;
+  sess_info.imsi       = imsi;
+  sess_info.Pdr_rules_ = session->get_all_pdr_rules();
+  if (!sess_info.Pdr_rules_.empty()) {
+    // Need to fix the error TODO
+    auto& rule        = sess_info.Pdr_rules_.front();
+    sess_info.ip_addr = rule.pdi().ue_ip_adr();
+  }
   sess_info.local_f_teid = session->get_local_teid();
   session->sess_infocopy(&sess_info);

```

Comment:
Doesn't `sess_infocopy` already put a bunch of information into SessionInfo? Why do we set some of these fields outside of this function and some inside?

Reply:
I need to move there.

Reply:
Adopted, moved all the code.